File: examples/csharp/Patterns/Observer.cs
namespace DesignPatterns.Observer
{
    public interface IObserver
    {
        void Update(float temperature, float humidity, float pressure);
    }

    public class WeatherStation
    {
        private readonly List<IObserver> _observers = new();
        private float _temperature, _humidity, _pressure;

        public void Register(IObserver observer) => _observers.Add(observer);
        public void Remove(IObserver observer) => _observers.Remove(observer);

        public void SetMeasurements(float temp, float humidity, float pressure)
        {
            _temperature = temp; _humidity = humidity; _pressure = pressure;
            Console.WriteLine($"  WeatherStation: new data ({temp}F, {humidity}%, {pressure}hPa)");
            foreach (var o in _observers) o.Update(temp, humidity, pressure);
        }
    }

    public class CurrentConditionsDisplay : IObserver
    {
        public void Update(float temperature, float humidity, float pressure) =>
            Console.WriteLine($"  CurrentConditions: {temperature}F, {humidity}% humidity");
    }

    public class ForecastDisplay : IObserver
    {
        private float _lastPressure = 1013.25f;

        public void Update(float temperature, float humidity, float pressure)
        {
            var forecast = pressure > _lastPressure ? "improving" :
                           pressure < _lastPressure ? "worsening" : "stable";
            Console.WriteLine($"  Forecast: Weather is {forecast}");
            _lastPressure = pressure;
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var station = new WeatherStation();
            station.Register(new CurrentConditionsDisplay());
            station.Register(new ForecastDisplay());

            station.SetMeasurements(75, 65, 1013.1f);
            Console.WriteLine();
            station.SetMeasurements(80, 70, 1015.2f);
        }
    }
}

File: examples/cpp/observer.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Observer {
public:
    virtual ~Observer() = default;
    virtual void update(float temp, float humidity) = 0;
};

class WeatherStation {
    std::vector<Observer*> observers_;
    float temperature_ = 0;
    float humidity_ = 0;
public:
    void addObserver(Observer* obs) { observers_.push_back(obs); }
    void setMeasurements(float temp, float humidity) {
        temperature_ = temp;
        humidity_ = humidity;
        notify();
    }
    void notify() {
        for (auto* obs : observers_) obs->update(temperature_, humidity_);
    }
};

class PhoneDisplay : public Observer {
public:
    void update(float temp, float humidity) override {
        std::cout << "  Phone Display - Temp: " << temp << "F, Humidity: " << humidity << "%" << std::endl;
    }
};

class WindowDisplay : public Observer {
public:
    void update(float temp, float humidity) override {
        std::cout << "  Window Display - Temp: " << temp << "F, Humidity: " << humidity << "%" << std::endl;
    }
};

int main() {
    std::cout << "=== Observer: Weather Station ===" << std::endl;

    WeatherStation station;
    PhoneDisplay phone;
    WindowDisplay window;

    station.addObserver(&phone);
    station.addObserver(&window);

    std::cout << "\nWeather update 1:" << std::endl;
    station.setMeasurements(72.5f, 65.0f);

    std::cout << "\nWeather update 2:" << std::endl;
    station.setMeasurements(80.0f, 90.0f);

    return 0;
}
